File: app/src/main/kt/estyle/teabaike/adapter/MainAdapter.java
package estyle.teabaike.adapter;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import estyle.teabaike.R;
import estyle.teabaike.bean.MainBean;
import estyle.teabaike.databinding.ItemMainBinding;

public class MainAdapter extends RecyclerView.Adapter implements View.OnClickListener {

    private static final int TYPE_ITEM = -1;

    private Context mContext;
    private List<MainBean.DataBean> mDatas;

    private View mEmptyView;
    private List<View> mHeaderList;
    private List<View> mFooterList;
    private int mHeaderCount;
    private int mFooterCount;

    private OnItemClickListener mOnItemClickListener;

    public MainAdapter(Context context) {
        this.mContext = context;
        mDatas = new ArrayList<>();
    }

    // 刷新数据
    public void refreshDatas(List<MainBean.DataBean> datas) {
        this.mDatas.clear();
        addDatas(datas);

    }

    // 添加数据
    public void addDatas(List<MainBean.DataBean> datas) {
        this.mDatas.addAll(datas);
        notifyDataSetChanged();
    }

    // 添加头视图
    public void addHeaderView(View headerView) {
        if (mHeaderList == null) {
            mHeaderList = new ArrayList<>();
        }
        mHeaderList.add(headerView);
        mHeaderCount = mHeaderList.size();
    }

    // 添加w尾视图
    public void addFooterView(View footerView) {
        if (mFooterList == null) {
            mFooterList = new ArrayList<>();
        }
        mFooterList.add(footerView);
        mFooterCount = mFooterList.size();
    }

    // 设置空视图
    public void setEmptyView(View emptyView) {
        this.mEmptyView = emptyView;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder holder;
        if (viewType == TYPE_ITEM) {
            ItemMainBinding binding = DataBindingUtil.inflate(LayoutInflater.from(mContext),
                    R.layout.item_main,
                    parent,
                    false);
            View itemView = binding.getRoot();
            initItemViewListener(itemView);
            holder = new ItemViewHolder(itemView);
            ((ItemViewHolder) holder).setBinding(binding);
        } else {
            if (viewType < mHeaderCount) {
                View headerView = mHeaderList.get(viewType);
                holder = new HeaderViewHolder(headerView);
            } else {
                View footerView = mFooterList.get(viewType - mHeaderCount - mDatas.size());
                holder = new FooterViewHolder(footerView);
            }
        }
        return holder;
    }

    private void initItemViewListener(View itemView) {
        itemView.setOnClickListener(this);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ItemViewHolder) {
            ItemViewHolder itemViewHolder = (ItemViewHolder) holder;
            MainBean.DataBean data = mDatas.get(position - mHeaderCount);
            itemViewHolder.getBinding().setBean(data);
        }
    }

    @Override
    public int getItemCount() {
        int itemCount = mHeaderCount + mDatas.size() + mFooterCount;
        if (mDatas.size() > 0) {
            if (mEmptyView != null) {
                mEmptyView.setVisibility(View.GONE);
            }
            if (mFooterCount > 0) {
                for (View footerView : mFooterList) {
                    footerView.setVisibility(View.VISIBLE);
                }
            }
        }
        return itemCount;
    }

    @Override
    public long getItemId(int position) {
        return Long.parseLong(mDatas.get(position - mHeaderCount).getId());
    }

    @Override
    public int getItemViewType(int position) {
        if (mHeaderCount > 0) {
            if (position < mHeaderCount) {
                return position;
            }
        }

        if (mFooterCount > 0) {
            if (position > (mHeaderCount + mDatas.size()) - 1) {
                return position;
            }
        }
        return TYPE_ITEM;
    }

    @Override
    public void onClick(View view) {
        int position;
        switch (view.getId()) {
            default:
                if (mOnItemClickListener != null) {
                    position = ((RecyclerView) view.getParent()).getChildLayoutPosition(view);
                    mOnItemClickListener.onItemClick(position);
                }
                break;
        }
    }

    private static class HeaderViewHolder extends RecyclerView.ViewHolder {

        private HeaderViewHolder(View itemView) {
            super(itemView);
        }
    }

    private static class ItemViewHolder extends RecyclerView.ViewHolder {

        private ItemMainBinding binding;

        private ItemViewHolder(View itemView) {
            super(itemView);
        }

        public ItemMainBinding getBinding() {
            return binding;
        }

        public void setBinding(ItemMainBinding binding) {
            this.binding = binding;
        }
    }

    private static class FooterViewHolder extends RecyclerView.ViewHolder {

        private FooterViewHolder(View itemView) {
            super(itemView);
        }
    }

    public interface OnItemClickListener {
        void onItemClick(int position);
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.mOnItemClickListener = onItemClickListener;
    }

}

File: app/src/main/kt/estyle/teabaike/adapter/MainAdapter.kt
package estyle.teabaike.adapter

import android.view.ViewGroup
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.paging.PagedList
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import estyle.teabaike.R
import estyle.teabaike.adapter.viewholder.BaseViewHolder
import estyle.teabaike.databinding.ItemMainBinding
import estyle.teabaike.databinding.ViewHeadlineBinding
import estyle.teabaike.entity.HeadlineEntity
import estyle.teabaike.entity.MainEntity
import estyle.teabaike.util.ImageUtil
import kotlinx.android.synthetic.main.view_headline.view.*

class MainAdapter :
    BasePagedListAdapter<MainEntity.DataEntity, RecyclerView.ViewHolder>(diffCallback) {

    var itemClickCallback: ((id: Long) -> Unit)? = null
    private var headerCount: Int = 0

    lateinit var onHeadlineViewHolderCreatedCallback: ((holder: MainAdapter.HeadlineViewHolder) -> Unit)

    var headlineList: List<HeadlineEntity.DataEntity>? = null
        set(value) {
            value ?: return
            if (field != null) headerCount--
            field = value
            headerCount++
            notifyDataSetChanged()
        }

    override fun getItemCount(): Int = super.getItemCount() + headerCount

    override fun getItemId(position: Int): Long =
        if (position < headerCount)
            super.getItemId(position)
        else
            getItem(position - headerCount)!!.id!!.toLong()

    override fun getItemViewType(position: Int): Int {
        var viewType = super.getItemViewType(position)
        if (position < headerCount) {
            if (position == 0) viewType = TYPE_HEADER_HEADLINE
        }
        return viewType
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        lateinit var holder: RecyclerView.ViewHolder
        when (viewType) {
            TYPE_HEADER_HEADLINE -> {
                holder = HeadlineViewHolder(parent, R.layout.view_headline).apply { bind(headlineList!!) }
                onHeadlineViewHolderCreatedCallback(holder)
            }
            else -> holder = ViewHolder(parent, R.layout.item_main)
        }
        return holder
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is ViewHolder) {
            holder.bind(getItem(position - headerCount)!!)
        }
    }

    override fun onViewAttachedToWindow(holder: RecyclerView.ViewHolder) {
        super.onViewAttachedToWindow(holder)
        if (holder is HeadlineViewHolder) {
            holder.start()
        }
    }

    override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
        super.onViewDetachedFromWindow(holder)
        if (holder is HeadlineViewHolder) {
            holder.stop()
        }
    }

    override fun onCurrentListChanged(
        previousList: PagedList<MainEntity.DataEntity>?,
        currentList: PagedList<MainEntity.DataEntity>?
    ) {
        super.onCurrentListChanged(previousList, currentList)
        // todo 首页itemcount总是0
//        ZYLog.e("pre : " + previousList?.size + ", curr: " + currentList?.size + ", itemcount: " + itemCount + ", empty visible: " + emptyView?.visibility)
    }

    // todo 先出现item数据，头视图数据后出现，使用notify，位置向下滑
    // todo 讲道理，生命周期应该绑定到BannerView上，坐等ViewPager2吧
     inner class HeadlineViewHolder(parent: ViewGroup, layoutId: Int) :
        BaseViewHolder<List<HeadlineEntity.DataEntity>, ViewHeadlineBinding>(
            parent,
            layoutId
        ), DefaultLifecycleObserver {

        // todo 如何优雅地使用databinding
        override fun bind(currentItem: List<HeadlineEntity.DataEntity>) {
            itemView.dots_view.dotCount = currentItem.size
            itemView.headline_text_view.text = currentItem[0].title
            itemView.banner_view_old.setOnBannerSelectedListener { position ->
                itemView.headline_text_view.text = currentItem[position].title
                itemView.dots_view.selectedPosition = position
            }
            itemView.banner_view_old.setOnBannerClickListener { position ->
                this@MainAdapter.itemClickCallback?.invoke(currentItem[position].id!!.toLong())// todo  id is nonnull
            }
            val imageList = arrayListOf<String>().apply { currentItem.forEach { add(it.image!!) } }
            itemView.banner_view_old.loadImages(imageList) { path, view -> ImageUtil.show(view, path) }
        }

        fun start() {
            itemView.banner_view_old.isCircling = true
        }

        fun stop() {
            itemView.banner_view_old.isCircling = false
        }

        override fun onStart(owner: LifecycleOwner) {
            start()
        }

        override fun onStop(owner: LifecycleOwner) {
            stop()
        }

        override fun onDestroy(owner: LifecycleOwner) {
            itemView.banner_view_old.onDestroy()
        }
    }

    private inner class ViewHolder(parent: ViewGroup, layoutId: Int) :
        BaseViewHolder<MainEntity.DataEntity, ItemMainBinding>(parent, layoutId) {

        override fun create() {
            super.create()
            itemView.setOnClickListener {
                this@MainAdapter.itemClickCallback?.invoke(getItemId(adapterPosition))
            }
        }

        override fun bind(currentItem: MainEntity.DataEntity) {
            binding.main = currentItem
        }
    }

    // RecyclerView附属的比较新旧item差异的回调
    private class ItemCallback : DiffUtil.ItemCallback<MainEntity.DataEntity>() {
        override fun areItemsTheSame(
            oldItem: MainEntity.DataEntity,
            newItem: MainEntity.DataEntity
        ): Boolean = oldItem === newItem

        override fun areContentsTheSame(
            oldItem: MainEntity.DataEntity,
            newItem: MainEntity.DataEntity
        ): Boolean = oldItem == newItem
    }

    companion object {
        private const val TYPE_HEADER_HEADLINE = 1
        private val diffCallback = ItemCallback()
    }
}